include/revng/Pipeline/Invokable.h

```diff
@@ -38,70 +118,264 @@ concept HasName = requires() {
 /// * a void run(T...) method where the first argument must be a Context and
 ///   every other type must be the most derived type of a Container.
 template<typename InvokableType, typename FirstRunArg, typename... Rest>
-concept Invokable = (derived_from<std::decay_t<Rest>, ContainerBase> and ...)
+concept Invokable = (IsContainer<Rest> and ...)
                     and convertible_to<Context &, std::remove_cv_t<FirstRunArg>>
                     and HasName<InvokableType>;
 
 // clang-format on
 
-namespace detail {
-using StringArrayRef = llvm::ArrayRef<std::string>;
+template<typename T>
+class ArgumentInfo {
+public:
+  constexpr ArgumentInfo(const char *Name, T Default) :
+    Default(Default), Name(Name) {}
+
+  using Type = T;
+  T Default;
+  const char *Name;
+};
 
 template<typename T>
-auto &getContainerFromName(ContainerSet &Containers, llvm::StringRef Name) {
-  return llvm::cast<T>(Containers[Name]);
-}
+concept HasArgumentsInfo = requires(T) {
+  { T::ArgumentInfos };
+};
 
-template<typename InvokableType, typename... Args, size_t... S>
-void invokeImpl(Context &Ctx,
-                InvokableType &Pipe,
-                void (InvokableType::*F)(Context &, Args...),
-                ContainerSet &Containers,
-                const StringArrayRef &ArgsNames,
-                const std::integer_sequence<size_t, S...> &) {
-
-  using namespace std;
-  (Pipe.*F)(Ctx,
-            getContainerFromName<decay_t<Args>>(Containers, ArgsNames[S])...);
-}
+namespace detail {
 
-template<typename InvokableType, typename... Args, size_t... S>
-void invokeImpl(const Context &Ctx,
-                InvokableType &Pipe,
-                void (InvokableType::*F)(const Context &, Args...),
-                ContainerSet &Containers,
-                const StringArrayRef &ArgsNames,
-                const std::integer_sequence<size_t, S...> &) {
-  using namespace std;
-  (Pipe.*F)(Ctx,
-            getContainerFromName<decay_t<Args>>(Containers, ArgsNames[S])...);
-}
+  template<typename Invokable, size_t Index>
```

Comment:
Why is this indented?
`clang-format` is behaving due to concepts.

Reply:
i have not idea, if deindent it, it gets indented again

Reply:
It's because of the `concept`. We use [this trick](https://github.com/revng/revng/blob/develop/include/revng/ADT/Concepts.h#L37) elsewhere.